lib/server.js

```diff
@@ -1352,7 +1352,10 @@ Server.prototype._finishReqResCycle = function _finishReqResCycle(
     var self = this;
     var route = req.route; // can be undefined when 404 or error
 
-    if (res._finished) {
+    // if the returned err value was a string, then we're handling the
+    // next('foo') case where we redirect to another middleware stack. don't
+    // do anything here because we're not done yet.
+    if (_.isString(err)) {
```

Comment:
Probably this check should not happen in the `finishReqResCycle` method itself, but this method is called from multiple places so rather than gate the check we'll just bail here. I think the old check here worked because we didn't have to handle this use case, but now we have to be more explicit.